### PR TITLE
Add in-place `_foreach_copy`

### LINKED ISSUE #107162: Eager foreach_copy
### 🚀 The feature, motivation and pitch

It would be useful to have a foreach_copy in eager! For example, in rprop, there is a for loop copying grads to prevs that could be foreach'd once the API exists.

A GitHub search shows that this has been attempted before and can be revived again https://github.com/pytorch/pytorch/pull/47072. 

cc'ing @crcrpar @jiayisuse 

### Alternatives

Slowness

### Additional context

_No response_

cc @vincentqb @jbschlosser @albanD @crcrpar @mcarilli

### PR DESCRIPTION
**Pull request by crcrpar**
Fixes #107162

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

File: torch/optim/rprop.py [hunk 1]
```diff
@@ -300,11 +300,10 @@ def _view_complex_as_real(tensor_list):
...
         # to keep referring to the buffer as grouped_grads.
-        for i in range(len(grouped_prevs)):
-            if maximize:
-                grouped_prevs[i].copy_(grouped_grads[i] * -1)
-            else:
-                grouped_prevs[i].copy_(grouped_grads[i])
+        if maximize:
+            torch._foreach_copy_(grouped_prevs, torch._foreach_neg(grouped_grads))
+        else:
+            torch._foreach_copy_(grouped_prevs, grouped_grads)
```
Comment (janeyx99) [inline torch/optim/rprop.py:306]:
```suggestion
        torch._foreach_copy_(grouped_prevs, grouped_grads)
        if maximize:
            torch._foreach_neg_(grouped_prevs)            
```
to save on intermediate memory

File: torch/testing/_internal/opinfo/core.py [hunk 2]
```diff
@@ -2687,13 +2687,13 @@ def __init__(
             torch_ref_method,
             torch_ref_inplace,
         ) = get_foreach_method_names(name)
-        if name == "zero":
+        if name in {"zero", "copy"}:
```
Comment (janeyx99) [inline torch/testing/_internal/opinfo/core.py:2690]: It feels like we should have an attribute of the ForeachFuncInfo be `must_be_inplace` or `has_no_out_of_place_impl` (you can decide a name) and to use this attr throughout the tests for consolidation

File: test/test_foreach.py [hunk 3]
```diff
@@ -1075,6 +1092,27 @@ def test_0dim_tensor_overload_exception(self):
...
+                for t, s in zip(ref_input, sample.args[0]):
+                    copy_(t, s, non_blocking)
+                self.assertEqual(sample.input, ref_input)
+                if torch.cuda.device_count() > 1:
+                    device = torch.device("cuda", 1)
+                    rhs_tensors = [t.to(device) for t in sample.args[0]]
+                    foreach_copy_(sample.input, rhs_tensors, non_blocking)
+                    for t, s in zip(ref_input, rhs_tensors):
+                        copy_(t, s, non_blocking)
+                    self.assertEqual(sample.input, ref_input)
```
Comment (janeyx99) [inline test/test_foreach.py:1114]:
```suggestion
                    self.assertEqual(ref_input, sample.input)
```
it's expected, actual for convention, right?

sam in line 1107

File: test/test_foreach.py [hunk 4]
```diff
@@ -902,7 +919,7 @@ def test_inplace_foreach_leaf_check_and_grad_fn(self, device, dtype, op):
         dtypes=(torch.float,),
     )
     def test_outplace_with_invalid_grads(self, device, dtype, op):
-        if op.name in {"_foreach_zero"}:
+        if op.name in {"_foreach_zero", "_foreach_copy"}:
```
Comment (janeyx99) [inline test/test_foreach.py:922]: Like I mentioned below, we can make this an attribute

File: aten/src/ATen/native/cuda/ForeachBinaryOpList.cu [hunk 5]
```diff
@@ -248,4 +250,44 @@ FOREACH_BINARY_OP_LIST(
...
+  std::vector<std::vector<at::Tensor>> tensor_lists{src.vec(), self.vec()};
+
+  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(
+      ScalarType::Half,
+      ScalarType::BFloat16,
+      ScalarType::Bool,
+      self[0].scalar_type(),
+      "foreach_tensor_copy",
+      [&]() {
+        using opmath_t = at::opmath_type<scalar_t>;
```
Comment (janeyx99) [inline aten/src/ATen/native/cuda/ForeachBinaryOpList.cu:280]: This is for my understanding: I am not sure I wrapped my head correctly here. From what it looks like, it looks like we put the src and the thing we're copying from into a list of lists of Tensors...and then we call Identity on all of them with...a UNARY op? Shouldn't this be a binary op?

**Review by janeyx99 — COMMENTED** [review]
Wow you are so speedy! Left some nits and qs, overall looks good

File: test/test_foreach.py (hunk 3 above)
Comment (crcrpar) [reply test/test_foreach.py:1114]: Is that torch.testing.assert_close's arguments? unittest.TestCase.assertEqual has x and y https://docs.python.org/3/library/unittest.html

File: aten/src/ATen/native/cuda/ForeachBinaryOpList.cu (hunk 5 above)
Comment (crcrpar) [reply aten/src/ATen/native/cuda/ForeachBinaryOpList.cu:280]: Binary op functors take two or three tensorlists and require the given operator to take two arguments. The signature of inplace foreach copy does look binary bit the operator it calls is unary as it takes one argument and forwards it.

**Comment by vadimkantorov** [comment]
Will also fix https://github.com/pytorch/pytorch/issues/90539

**Review by janeyx99 — APPROVED** [review]
thanks!

**Pull request closed without merging**